xla/service/gpu/gpu_compiler_test.cc

```diff
@@ -1056,67 +1056,75 @@ ENTRY main {
             expect_custom_kernel_fusion_rewriter_has_run);
 }
 
-struct PassRunIndex {
-  int first_run = std::numeric_limits<int>::max();
-  int last_run = std::numeric_limits<int>::min();
-};
-
-// Checks that both passes have actually run and that the first run of the
-// `after` pass is after the last run of the `before` pass.
-void VerifyPassOrder(
-    const absl::flat_hash_map<std::string, PassRunIndex>& passes,
-    absl::string_view before, absl::string_view after) {
-  ASSERT_TRUE(passes.contains(before))
-      << "Expected pass did not run: " << before;
-  ASSERT_TRUE(passes.contains(after)) << "Expected pass did not run: " << after;
-  EXPECT_LT(passes.at(before).last_run, passes.at(after).first_run)
-      << "Pass " << before << " ran after " << after;
-}
-
-// Traverses the module's pass metadata and gathers, for each pass, its smallest
-// and largest run index.  If a pass p0's run index is smaller than another pass
-// p1's run index, then p0 ran before p1.
-absl::flat_hash_map<std::string, PassRunIndex> GatherPassOrderInformation(
-    const HloModule& module) {
-  // Maps a pass name to its first and last index.
-  absl::flat_hash_map<std::string, PassRunIndex> passes;
-  int run_index = 0;
-  for (const HloPassMetadata& pass_metadata :
-       module.metadata().proto().pass_metadata()) {
-    auto& pass = passes[pass_metadata.pass_name()];
-    pass.first_run = std::min(pass.first_run, run_index);
-    pass.last_run = std::max(pass.last_run, run_index);
-    ++run_index;
+class PassOrderTest : public GpuCompilerTest {
+ public:
+  void SetDebugOptions(const DebugOptions& options) {
+    HloModuleConfig config = GetModuleConfigForTest();
+    config.set_debug_options(options);
+    SetModuleConfig(config);
   }
 
-  return passes;
-}
-
-TEST_F(GpuCompilerPassTest, PassesAreRunInCorrectOrder) {
-  constexpr absl::string_view constant_module = R"(
+  void SetModuleConfig(const HloModuleConfig& config) {
+    constexpr absl::string_view constant_module = R"(
 ENTRY main {
   ROOT constant = f32[] constant(0)
 })";
+    TF_ASSERT_OK_AND_ASSIGN(
+        std::unique_ptr<VerifiedHloModule> module,
+        ParseAndReturnVerifiedModule(constant_module, config));
+    TF_ASSERT_OK_AND_ASSIGN(optimized_module_,
+                            GetOptimizedModule(std::move(module)));
+  }
 
-  TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<VerifiedHloModule> module,
-                          ParseAndReturnVerifiedModule(constant_module));
-  TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<HloModule> optimized_module,
-                          GetOptimizedModule(std::move(module)));
+  // Fails if any of the passes with names matching the regular expression
+  // first_regex run after any of the passes matching last_regex or if none of
+  // the executed passes matches first_regex or last_regex.
+  void VerifyPassOrder(absl::string_view first_regex,
+                       absl::string_view last_regex) {
```

Comment:
I'm having a bit of a hard time following the terminology between `first_latest` and `last_earliest`.

Could we stick to the previous terminology? (I.e. `first` -> `before`, `earliest` -> `first`, ...)
We could even make it `before_pass_regex` to enhance clarity further.